### ISSUE TITLE
Rewrite/Rethought of some stuff

### ISSUE THREAD

**Issue by C9Glax** [C9Glax/tranga#41]
**tl;dr: No new features as long as this issue is open, only major bugs will be fixed.**

So at the moment I have the following problem:
All the parts that make this project work, are somewhat intertwined, despite my efforts to seperate them.
This makes debugging stuff very hard, and issues crop up where I might not expect them to be.

The only part that seems to constantly work, and is easy to debug, is the actual download/connector aspect.
Because this part is actually seperated and runs fully independently through  interfaces and only does the stuff it is supposed to do: Download Manga.

So I have to do some major rethinking of parts of Tranga, which won't actually change how the end-user might see things, but will make it easier to identify where the actual problem is (example: Today I tried to add a new Manga, but got stuck on the actual search. I had to look at every component, from website, to api, to the taskmanager, connector, etc. to figure out where the issue was).

So here is my idea:

- Use the parts of the Connectors, because they actually work and have been working (with minor edge-cases) flawlessly for now.
- Independently from that create an interface that let's you interact with the connectors (select Manga, select Chapters, etc.)
- Create a new Job-System (rename that from Task, because that has all kinds of complications I have figured), that calls this connector-interface.
- Create and interface for this job-system.
- And finally take it all together and write an interface/API that let's you interact with both the connector-interface and the job-system.

This is all already somewhat how things work, but there are bits and pieces at wrong places and the boundaries are kind of blurry, so there has to be a complete rewrite of the job-system (this part is such a headache), and clear separation between components.

**In the meantime, there won't be new features (current requests will probably influence some of the rewrite), and only major bugs will be fixed.**

**Comment by C9Glax** [comment]
Track progress [here](https://github.com/C9Glax/tranga/tree/41_-_trash_everything)
Warning: Barebones

**Comment by Xaneph** [comment]
Some thoughts on this:

1. What you are doing is great and I hope you feel encouraged
2. I hope part of the Rewrite includes better documentation. Many people are still trying to get this working thinking that is somewhat more functional than it is. 
3. is there any way to cancel a download task? when I start the container it resumes all DL jobs and never times out on jobs it will obviously never finish. 
4. Why can't we use the API key as the connector....it basically is the password.

**Comment by C9Glax** [comment]
Hello!
Lately I haven't had the time or motivation to work on this, but I have made some changes. For now most of the things work...
To answer your questions:

Yes I should write more clearly that this tool is not the do-all, be-all solution, and is just fetching images on a routine-basis. That would save me a lot of trouble. Thanks for pointing that out.
At the moment there is theoretically a way to cancel tasks, however the api does not implement that function... There should be a timeout on tasks (I think it's three minutes)...

What do you mean by your last question?

**Comment by C9Glax** [comment]
Progress Update:
The Framework is done. Tasks are now called Jobs, for better distinction from the System-Class Task.
Implemented an updated Progress-Tracking-System, that also allows for Job-Cancellation.
Now working on the API, which will be used by both the Web and CLI clients.

**Comment by C9Glax** [comment]
New Update:

- API is functional, not all paths tested yet
- Website is up and running

Currently working (API calls and Website):

- Search for new Manga
- Add Monitoring Jobs
- Get monitoring jobs

Improvements: You no longer need to mount imageCache to a static path, the API can now handle requests for those covers.

**Comment by C9Glax** [comment]
New Working:

- Start Jobs
- Cancel Jobs
- Delete Jobs
- Number of running and waiting jobs

Fixed some issues with stop/running program (loading `jobs.json`), now parentJobs get assigned their subJobs (maybe renamed to domJobs? >.> )

**Comment by C9Glax** [comment]
New Working:

- #32 You still have to select the connector however

Optimized some Parsing in Mangasee as well
Fixed the logging issue where multiple thread writing logs would crash the logger

**Comment by C9Glax** [comment]
Just merged into cuttingedge.

Caution: **BREAKING CHANGES**
There is no migration, backup your existing settings, configs, etc.

You can (at the moment) switch back to master in case cuttingedge doesn't work, with the old configs.

**Comment by vodkapmp** [comment]
Hi!
With the new cuttingedge, I can see that the mount for `/usr/share/Tranga-API` isn't there anymore.. 
How do jobs persist through updates and such when there is no persistent storage for anything but the mangas themselves?
Surely there's a mount point for whatever controls the storage of what manga you've got added to Tranga? Would be, uh quite inconvenient if I have to readd everything on every update..

**Comment by C9Glax** [comment]
True!
I guess I removed one line to many, but I will add that back

**Comment by C9Glax** [comment]
Complete with #57
